tests/test_knn_functions.cpp

```diff
@@ -0,0 +1,165 @@
+/**
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+#include <gflags/gflags_declare.h>
+#include <gtest/gtest.h>
+#include <algorithm>
+#include <array>
+#include <cstdint>
+#include <memory>
+#include <random>
+
+#include "faiss/utils/distances.h"
+
+static const unsigned NPTS{1000};
+static const unsigned DIM{128};
+static const std::array<unsigned, 3> TOP_K_LIST{1, 50, 100};
+static const std::array<unsigned, 2> NUMS_QUREY_LIST{10, 30};
+
+using VectorDataPtr = std::unique_ptr<float[]>;
+
+VectorDataPtr generateRandomVector(
+        const unsigned npts,
+        const unsigned dim,
+        int seed = 0) {
+    std::mt19937 gen(seed);
+    std::uniform_real_distribution<> dis(-1.0, 1.0);
+    VectorDataPtr data = std::make_unique<float[]>(npts * dim);
+    for (unsigned n{0}; n < npts; ++n) {
+        for (unsigned d = 0; d < dim; ++d) {
+            data[n * dim + d] = dis(gen);
+        }
+    }
+    return std::move(data);
+}
+
+float computeL2SqrDist(
+        const float* vec1,
+        const float* vec2,
+        const unsigned dim) {
+    float dist = 0;
+    for (unsigned i{0}; i < dim; ++i) {
+        float diff = vec1[i] - vec2[i];
+        dist += diff * diff;
+    }
+    return dist;
+}
+
+float computeInnerProductDist(
+        const float* vec1,
+        const float* vec2,
+        const unsigned dim) {
+    float dist = 0;
+    for (unsigned i{0}; i < dim; ++i) {
+        dist += vec1[i] * vec2[i];
+    }
+    return dist;
+}
+
+void checkKNNResults(
+        const float* queries,
+        const float* db,
+        size_t nq,
+        size_t ndb,
+        size_t dim,
+        size_t k,
+        const float* distances,
+        const int64_t* indexes,
+        bool is_l2 = true) {
+    for (size_t q = 0; q < nq; ++q) {
+        std::vector<std::pair<float, int64_t>> res_pairs;
+        for (int64_t i = 0; i < ndb; ++i) {
+            float dist{0};
+            if (is_l2) {
+                dist = computeL2SqrDist(queries + q * dim, db + i * dim, dim);
+            } else {
+                dist = computeInnerProductDist(
+                        queries + q * dim, db + i * dim, dim);
+            }
+            res_pairs.emplace_back(dist, i);
+        }
+        std::sort(
+                res_pairs.begin(),
+                res_pairs.end(),
+                [is_l2](const std::pair<float, int64_t>& a,
+                        const std::pair<float, int64_t>& b) {
+                    return is_l2 ? a.first < b.first : a.first > b.first;
+                });
+        for (size_t j = 0; j < k; ++j) {
+            ASSERT_EQ(res_pairs[j].second, indexes[q * k + j]);
+            // WARNING: There will be a very small fractional error :(
+            // ASSERT_FLOAT_EQ(res_pairs[j].first, distances[q * k + j]);
```

Comment:
Will this be a flaky test?

Reply:
I rewrote the test, but the results appear to be in error

```cpp
            ASSERT_EQ(res_pairs[j].second, indexes[q * k + j]);
            // WARNING: There will be a very small fractional error :(
            // ASSERT_FLOAT_EQ(res_pairs[j].first, distances[q * k + j]);
            if (is_l2) {
                ASSERT_FLOAT_EQ(
                        faiss::fvec_L2sqr(
                                queries + q * dim,
                                db + res_pairs[j].second * dim,
                                dim),
                        distances[q * k + j]);
            }
```

but i got error follow 
```
Test project /home/byte/Code/ANNS/faiss/build
    Start 115: TestKNNFunctions.knn_L2sqr
1/1 Test #115: TestKNNFunctions.knn_L2sqr .......***Failed    0.02 sec
WARNING clustering 1000 points to 40 centroids: please provide at least 1560 training points
Running main() from /home/byte/Code/ANNS/faiss/build/_deps/googletest-src/googletest/src/gtest_main.cc
Note: Google Test filter = TestKNNFunctions.knn_L2sqr
[==========] Running 1 test from 1 test suite.
[----------] Global test environment set-up.
[----------] 1 test from TestKNNFunctions
[ RUN      ] TestKNNFunctions.knn_L2sqr
/home/byte/Code/ANNS/faiss/tests/test_knn_functions.cpp:105: Failure
Expected equality of these values:
  faiss::fvec_L2sqr( queries + q * dim, db + res_pairs[j].second * dim, dim)
    Which is: 0
  distances[q * k + j]
    Which is: 2.2888184e-05
/home/byte/Code/ANNS/faiss/tests/test_knn_functions.cpp:105: Failure
Expected equality of these values:
  faiss::fvec_L2sqr( queries + q * dim, db + res_pairs[j].second * dim, dim)
    Which is: 0
  distances[q * k + j]
    Which is: 2.2888184e-05
/home/byte/Code/ANNS/faiss/tests/test_knn_functions.cpp:105: Failure
Expected equality of these values:
  faiss::fvec_L2sqr( queries + q * dim, db + res_pairs[j].second * dim, dim)
    Which is: 0
  distances[q * k + j]
    Which is: 2.2888184e-05
[  FAILED  ] TestKNNFunctions.knn_L2sqr (4 ms)
[----------] 1 test from TestKNNFunctions (4 ms total)

[----------] Global test environment tear-down
[==========] 1 test from 1 test suite ran. (4 ms total)
[  PASSED  ] 0 tests.
[  FAILED  ] 1 test, listed below:
[  FAILED  ] TestKNNFunctions.knn_L2sqr

 1 FAILED TEST


0% tests passed, 1 tests failed out of 1

Total Test time (real) =   0.02 sec

The following tests FAILED:
        115 - TestKNNFunctions.knn_L2sqr (Failed)
Errors while running CTest
```
that means even though the two points have the correct nearest neighbor relationship, the knn function's distance calculation is subject to error

1. **Is it possible to get an error with the knn function in the library?**
2. **Or there might be a problem with the generated random data in my test?**